aten/src/ATen/native/UnaryOps.cpp

```diff
@@ -188,7 +188,15 @@ Tensor imag(const Tensor& self) {
 }
 
 Tensor& conj_out(Tensor& result, const Tensor& self) { return unary_op_impl_out(result, self, conj_stub); }
-Tensor conj(const Tensor& self) { return unary_op_impl(self, at::conj_out); }
+
+Tensor _conj(const Tensor& self) { return unary_op_impl(self, at::conj_out); }
+
+Tensor conj(const Tensor& self) {
+  if (!self.is_complex()) {
+    return self;
+  }
+  return at::_conj(self);
+}
```

Comment:
This seems BC-breaking: previously if a user passed in a real tensor to conj, they'd get back a copy of said tensor. What was the rationale behind this change?

FWIW, in numpy, conj returns a copy of the original tensor if it wasn't complex:
```
>>> x = np.array([0.1, 0.1])
>>> y = np.conj(x)
>>> y.flags
...
OWNDATA : True
```

Reply:
As discussed in this issue: #41857 switching to the tensorflow style gradients would involve addition of a bunch of `conj` in the gradient definitions. In order to preserve the backward performance for real tensors, it seems like a good option to make `torch.conj` a no op for real tensors.

Reply:
I see, so the decision in that thread was to go for TF-style gradients. However, TF-style gradients require conj-calls everywhere in the gradient formulas. Because of that, we're trying to optimize calls to conj to avoid regressions.

The downside of calling the operator "fast_conj" is that writers of gradient formulas might use "conj" instead of "fast_conj". I think it is probably OK for us to break BC and to diverge from numpy and have our "conj" not do copies when possible but I am not sure which of these two solutions would be better.

Exploring the design space some more, I noticed that there were some comments about having a conj flag on tensors that reinterprets the tensor as its conjugate (https://github.com/pytorch/pytorch/issues/41857#issuecomment-665129647). Was there a decision on that? That seems like it would eliminate the need to have a `fast_conj`

Reply:
I understand why we want to have a function with this behavior, but have a few questions/suggestions:

- I would prefer our `conj` have the same behavior as NumPy's
- `torch.conj` is, conceptually, a universal function, and all universal functions create new outputs
  - for example, `torch.abs(uint8)` produces a new tensor, even though it could mathematically fulfill its contract by returning self
  - analogously, would we expect `torch.ceil(int64)` to return self or a copy of self? 
- do we expect this function to be used by anyone but developers? Is making a non-user-facing function `self_or_conj` an option to address the scenarios we'd like to target? If we're uncertain about those scenarios, could we start with this?

I appreciate @zou3519's concern about gradient writers mistakenly using `conj` instead of `self_or_conj`, and I agree that's a real possibility, but if we're consistent about using the latter and do our job in code review I think it'll be OK. I'm also less concerned about introducing a fixable performance issue vs. creating an inconsistent and semantically confusing UX for end users.

Reply:
Although I am leaning towards "faithfully replicate numpy behavior" (@anjali411, I think I mentioned that I was having trouble deciding between these two choices in our call), I want to point out some reasons why fast conj by default might make sense.

> do we expect this function to be used by anyone but developers? Is making a non-user-facing function self_or_conj an option to address the scenarios we'd like to target? If we're uncertain about those scenarios, could we start with this?

`fast_conj` absolutely would be used by non-developers. Any time you write code that you want to generalize over both reals and complex you will write conjugations to handle the complex case, and if they are not no-ops, you will unnecessarily slow down the execution of your code in the real case. Some more support from boedekker here https://github.com/pytorch/pytorch/issues/41857#issuecomment-663152126

> Exploring the design space some more, I noticed that there were some comments about having a conj flag on tensors that reinterprets the tensor as its conjugate (#41857 (comment)). Was there a decision on that?

For the BC consideration here, it doesn't help. If we have a fancy version of conjugate that returns a view rather than new tensor, that would still be incompatible with the Numpy behavior. So if you want to keep numpy behavior, you're still going to have to come up with a new name for `view_conj` or whatever.

Reply:
cc @boeddeker

Reply:
I would lean towards @ezyang opinion here personally. In particular, "fast_conj absolutely would be used by non-developers." and not having it would mean that people will start adding a lot of `if t.is_complex()` in their code.

@mruberry are there functions for which we explicitly plan to have a different API compared to numpy? And maybe provide a numpy-compliant version in a different namespace? Could this function be one of them?

Reply:
I am working a lot with numpy and I must say, I never checked if numpy.conj returns a copy or hold the original data.
Sometimes I got the impression, that it is a no op, also for complex numbers, when I measured the execution time (execution time is close to nothing compared to any other function).

Nevertheless, assuming that conj returns a copy in the source tensor, also for real tensors, is a bad idea.

When you introduce `fast_conj`, I would never use `conj`.

In context of torch, I would prefer, when conj never creates a copy.
While the calculation of the copy may be suboptimal, the increased memory consumption is critical.
In numpy you can calculate on demand the conjugate and it will be deleted a few moments later.
In torch, it may be kept for the backward graph. When you call multiple times conj, you store several copies.

So maybe it would also be worth considering, that the conjugate operation gets an argument to enable something like `torch.utils.checkpoint.checkpoint` or have the `conj` flag, that was earlier mentioned.

Reply:
cc @zdevito this may be one of the first cases where we have existentially important laziness on tensors

Reply:
> @mruberry are there functions for which we explicitly plan to have a different API compared to numpy? And maybe provide a numpy-compliant version in a different namespace? 

Yes, but we should strive to limit these discrepancies. In particular we want a compelling reason for the difference (the NumPy behavior would be inconsistent with PyTorch, or it would require an extremely painful deprecation, ...). 

> Could this function be one of them?

Sure. I'm actually less worried about the NumPy compatibility of this function than the PyTorch inconsistency of it. It seems like we're trying to combine the ideas of `torch.conj()` the unary ufunc and a `.conj` attribute with a corresponding view function, like `.real`, `.imag`, and `.T`.